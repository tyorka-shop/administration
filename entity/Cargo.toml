[package]
edition = "2021"
name = "entity"
version = "0.1.0"

[lib]
name = "entity"
path = "src/lib.rs"

[dependencies]
eyre = "0.6.8"
macros = { path = "../macros" }
md5 = "0.7.0"
reqwest = { version = "0.11.11", features = ["json"] }
serde = { version = "1.0.144", features = ["serde_derive"] }
sqlx = { version = "0.6", features = [ "runtime-tokio-rustls", "sqlite" ] }
tokio = "1.20.1"
log = "0.4.17"