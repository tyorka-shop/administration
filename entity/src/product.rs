use sqlx::SqlitePool;

#[derive(macros::Entity)]
#[table_name = "products"]
pub struct Product {
    pub id: String,
    pub cover_id: String,
    pub title_en: String,
    pub title_ru: String,
    pub description_en: String,
    pub description_ru: String,
    pub price: Option<i64>,
    pub show_in_gallery: bool,
    pub show_in_shop: bool,
    pub created_at: chrono::NaiveDateTime,
    pub updated_at: chrono::NaiveDateTime,
}

impl Product {
    pub fn new_fixture() -> Self {
        Self {
            id: "07d7b72c-5b2e-4a35-a257-158496993dcc".into(),
            cover_id: format!("{:x}", md5::compute("cover_id")),
            title_en: "title".to_string(),
            title_ru: "заголовок".to_string(),
            description_en: "description".to_string(),
            description_ru: "описание".to_string(),
            price: None,
            show_in_gallery: true,
            show_in_shop: false,
            created_at: chrono::NaiveDateTime::from_timestamp(0, 0),
            updated_at: chrono::NaiveDateTime::from_timestamp(0, 0),
        }
    }

    pub fn new_fixture_with_id(id: &str) -> Self {
        let mut result = Self::new_fixture();
        result.id = id.into();
        result
    }

    pub async fn get_gallery(db: &SqlitePool) -> Result<Vec<Product>, sqlx::Error> {
        let products = sqlx::query_as!(
            Product,
            r#"
            SELECT `products`.* 
            FROM `products` 
            left join `entity_order` 
                on `entity_order`.`entity_id` = `products`.`id` and `entity_order`.`type` = "gallery"
            WHERE `products`.`show_in_gallery` = 1
            order by `entity_order`.`idx` asc
            "#
        )
        .fetch_all(db)
        .await
        .unwrap();

        Ok(products)
    }

    pub async fn get_shop(db: &SqlitePool) -> Result<Vec<Product>, sqlx::Error> {
        let products = sqlx::query_as!(
            Product,
            r#"
            SELECT `products`.* 
            FROM `products` 
            left join `entity_order` 
                on `entity_order`.`entity_id` = `products`.`id` and `entity_order`.`type` = "shop"
            WHERE `products`.`show_in_shop` = 1
            order by `entity_order`.`idx` asc
            "#
        )
        .fetch_all(db)
        .await
        .unwrap();

        Ok(products)
    }

    pub async fn save_order(
        db: &SqlitePool,
        entity_type: &str,
        list: &Vec<String>,
    ) -> Result<(), sqlx::Error> {
        let mut tx = db.begin().await?;
        let entity_type = entity_type.to_string();
        sqlx::query!(r#"delete from entity_order where `type` = $1"#, entity_type)
            .execute(&mut tx)
            .await
            .unwrap();

        for (i, id) in list.iter().enumerate() {
            let index = i.to_string();
            let id = id.to_string();
            match sqlx::query!(
                r#"insert into `entity_order` (`entity_id`, `type`, `idx`) values ($1, $2, $3)"#,
                id,
                entity_type,
                index
            )
            .execute(&mut tx)
            .await
            {
                Err(e) => {
                    log::error!("{}", e);
                    tx.rollback().await?;
                    return Err(e);
                }
                _ => {}
            };
        }
        tx.commit().await?;

        Ok(())
    }

    pub async fn save_gallery_order(
        db: &SqlitePool,
        list: &Vec<String>,
    ) -> Result<(), sqlx::Error> {
        Self::save_order(db, "gallery", list).await
    }

    pub async fn save_shop_order(db: &SqlitePool, list: &Vec<String>) -> Result<(), sqlx::Error> {
        Self::save_order(db, "shop", list).await
    }
}
