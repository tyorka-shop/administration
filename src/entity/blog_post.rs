use async_graphql::{ID, SimpleObject};

#[derive(SimpleObject)]
pub struct BlogPost {
  pub id: ID,
  pub src: String,
  pub url: String,
  pub color: String
}


#[derive(macros::Entity)]
#[table_name = "blog"]
pub struct Entity {
  pub id: String,
  pub src: String,
  pub url: String,
  pub color: String
}

impl From<&BlogPost> for Entity {
  fn from(post: &BlogPost) -> Self {
    Self {
      id: post.id.to_string(),
      src: post.src.clone(),
      url: post.url.clone(),
      color: post.color.clone()
    }
  }
}