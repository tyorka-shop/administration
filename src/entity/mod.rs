mod picture_size;
pub mod crop;
pub mod picture;
pub mod product;
pub mod multi_lang;
pub mod user;

pub use picture_size::PictureSize;

