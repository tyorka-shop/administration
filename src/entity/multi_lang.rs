use async_graphql::{InputObject, SimpleObject};


#[derive(SimpleObject, InputObject)]
#[graphql(input_name = "MultiLangInput")] 
pub struct MultiLang {
  pub en: String,
  pub ru: String
}