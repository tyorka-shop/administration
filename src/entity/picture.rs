use async_graphql::ID;
use serde::Serialize;

#[derive(macros::Entity)]
#[table_name = "pictures"]
pub struct Entity {
    pub id: String,
    pub color: String,
    pub original_size_width: i64,
    pub original_size_height: i64,
    pub crop_anchor_x: f64,
    pub crop_anchor_y: f64,
    pub crop_factor: f64,
    pub product_id: Option<String>,
    pub idx: Option<i64>,
}

use super::{
    crop::{Crop, Point},
    picture_size::PictureSize,
};

#[derive(Debug, Serialize, async_graphql::SimpleObject)]
pub struct Picture {
    pub id: ID,
    pub color: String,
    pub original_size: PictureSize,
    pub crop: Crop,
    #[serde(skip_serializing)]
    #[graphql(skip)]
    pub product_id: Option<String>,
}

impl From<Entity> for Picture {
    fn from(row: Entity) -> Self {
        Self {
            id: ID::from(row.id),
            color: row.color,
            original_size: PictureSize {
                width: row.original_size_width,
                height: row.original_size_height,
            },
            crop: Crop {
                anchor: Point {
                    x: row.crop_anchor_x,
                    y: row.crop_anchor_y,
                },
                factor: row.crop_factor,
            },
            product_id: row.product_id,
        }
    }
}

impl Picture {
    pub fn new(filename: &str, width: i64, height: i64, dominant_color: &str) -> Self {
        Self {
            id: ID::from(filename),
            color: dominant_color.to_string(),
            original_size: PictureSize { width, height },
            crop: Crop::default_square(width as u32, height as u32),
            product_id: None,
        }
    }
}

impl From<&Picture> for Entity {
    fn from(pic: &Picture) -> Self {
        Self {
            id: pic.id.to_string(),
            color: pic.color.clone(),
            original_size_width: pic.original_size.width,
            original_size_height: pic.original_size.height,
            crop_anchor_x: pic.crop.anchor.x,
            crop_anchor_y: pic.crop.anchor.y,
            crop_factor: pic.crop.factor,
            product_id: pic.product_id.clone(),
            idx: None,
        }
    }
}

impl Picture {
    pub async fn get_by_product_id(db: &sqlx::SqlitePool, id: &ID) -> sqlx::Result<Vec<Self>> {
        let rows = sqlx::query_as!(Entity, "select * from pictures where product_id = ? order by `idx`", id.0)
            .fetch_all(db)
            .await
            .unwrap()
            .into_iter()
            .map(|row| row.into())
            .collect::<Vec<Self>>();

        Ok(rows)
    }
}



#[cfg(test)]
impl Entity {
    pub fn mock() -> Self {
        Self {
            id: "4e2d05fa-d79c-401c-8cdf-275eb2dccbae".into(),
            color: "#000000".into(),
            original_size_width: 100,
            original_size_height: 100,
            crop_anchor_x: 0.5,
            crop_anchor_y: 0.5,
            crop_factor: 1.0,
            product_id: None,
            idx: None,
        }
    }
}