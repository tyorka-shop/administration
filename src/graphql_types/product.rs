use super::{multi_lang::MultiLang, ProductState};
use async_graphql::ID;
use serde::Serialize;

#[derive(async_graphql::SimpleObject)]
#[graphql(complex)]
pub struct Product {
    pub id: ID,
    #[graphql(skip)]
    pub cover_id: String,
    pub title: MultiLang,
    pub show_in_gallery: bool,
    pub show_in_shop: bool,
    pub price: Option<i64>,
    pub description: MultiLang,
}

impl From<entity::Product> for Product {
    fn from(row: entity::Product) -> Self {
        Self {
            id: ID::from(&row.id),
            cover_id: row.cover_id.clone(),
            description: MultiLang {
                en: row.description_en.clone(),
                ru: row.description_ru.clone(),
            },
            price: row.price,
            show_in_gallery: row.show_in_gallery,
            show_in_shop: row.show_in_shop,
            title: MultiLang {
                en: row.title_en.clone(),
                ru: row.title_ru.clone(),
            },
        }
    }
}

#[derive(async_graphql::InputObject, Serialize, Clone)]
pub struct ProductInput {
    pub id: ID,
    pub state: ProductState,
    pub pictures: Vec<ID>,
    pub cover_id: ID,
    pub title: MultiLang,
    pub show_in_gallery: bool,
    pub show_in_shop: bool,
    pub price: Option<i64>,
    pub description: MultiLang,
}

impl From<&ProductInput> for entity::Product {
    fn from(input: &ProductInput) -> Self {
        Self {
            id: input.id.to_string(),
            title_en: input.title.en.clone(),
            title_ru: input.title.ru.clone(),
            description_en: input.description.en.clone(),
            description_ru: input.description.ru.clone(),
            price: input.price,
            show_in_gallery: input.show_in_gallery,
            show_in_shop: input.show_in_shop,
            cover_id: input.cover_id.to_string(),
        }
    }
}

#[cfg(test)]
impl ProductInput {
    pub fn new_fixture() -> Self {
        let pic_id = ID::from(format!("{:x}", md5::compute("cover_id")));
        Self {
            id: "07d7b72c-5b2e-4a35-a257-158496993dcc".into(),
            state: ProductState::Draft,
            pictures: vec![pic_id.clone()],
            cover_id: pic_id,
            title: MultiLang {
                en: "title".to_string(),
                ru: "заголовок".to_string(),
            },
            show_in_gallery: true,
            show_in_shop: false,
            price: None,
            description: MultiLang {
                en: "description".to_string(),
                ru: "описание".to_string(),
            },
        }
    }
}
