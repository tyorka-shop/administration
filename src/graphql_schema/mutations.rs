use async_graphql::{Context, Object, Result};
use sqlx::SqlitePool;

use crate::entity::product::{self, Product, ProductInput};

pub struct Mutations;

#[Object]
impl Mutations {
    async fn save_product<'a>(&self, ctx: &Context<'a>, product: ProductInput) -> Result<Product> {
        let db = ctx.data::<SqlitePool>().unwrap();

        let entity = product::Entity::from(&product);

        if product.pictures.len() == 0 {
            return Err("Product must have at least one picture".into());
        }

        if product.show_in_shop {
            match product.price {
                Some(value) if value <= 0 => {
                    return Err("Product price must be greater than 0".into())
                }
                None => return Err("Product price must be set".into()),
                _ => {}
            }
        }

        entity.insert_or_update(db).await?;

        for (i, picture_id) in product.pictures.iter().enumerate() {
            let index = i.to_string();
            let product_id = product.id.to_string();
            let picture_id = picture_id.to_string();
            sqlx::query!(
                "update pictures set product_id = $1, `idx` = $2 where id = $3",
                product_id,
                index,
                picture_id
            )
            .execute(db)
            .await?;
        }

        Ok(Product::from(&entity))
    }
}

#[cfg(test)]
mod save_product {
    use crate::entity::product::ProductInput;

    use super::Mutations;
    use async_graphql::{EmptySubscription, InputType, Object, Request, Result, Variables};
    use sqlx::{sqlite::SqlitePoolOptions, SqlitePool};

    struct Queries;

    #[Object]
    impl Queries {
        async fn status(&self) -> Result<String> {
            Ok("Ok".into())
        }
    }

    async fn setup_db() -> Result<SqlitePool> {
        let db = SqlitePoolOptions::new()
            .max_connections(1)
            .connect(":memory:")
            .await
            .unwrap();

        sqlx::migrate!().run(&db).await.unwrap();

        Ok(db)
    }

    fn make_request(product: &ProductInput) -> Request {
        let vars = Variables::from_json(serde_json::json!({ "product": product.to_value() }));

        let mutation = r#"mutation AddProduct($product: ProductInput!) { saveProduct(product: $product) { id } }"#;

        Request::new(mutation).variables(vars)
    }

    async fn request(product: &ProductInput) -> (async_graphql::Response, sqlx::SqlitePool) {
        let schema = async_graphql::Schema::build(Queries, Mutations, EmptySubscription).finish();
        let db = setup_db().await.unwrap();

        let request = make_request(product);

        (schema.execute(request.data(db.clone())).await, db)
    }

    #[tokio::test]
    pub async fn valid() {
        let product = ProductInput::mock();
        let (result, _) = request(&product).await;

        assert_eq!(result.errors.first(), None);

        insta::assert_json_snapshot!(result.data.into_json().unwrap());
    }

    #[tokio::test]
    pub async fn empty_pictures() {
        let mut product = ProductInput::mock();
        product.pictures = vec![];

        let (result, _) = request(&product).await;

        assert_eq!(
            result.errors.first().unwrap().message,
            "Product must have at least one picture"
        );
    }

    #[tokio::test]
    pub async fn price_must_be_set() {
        let mut product = ProductInput::mock();
        product.show_in_shop = true;
        product.price = None;

        let (result, _) = request(&product).await;

        assert_eq!(
            result.errors.first().unwrap().message,
            "Product price must be set"
        );
    }

    #[tokio::test]
    pub async fn price_must_be_gt_0() {
        let mut product = ProductInput::mock();
        product.show_in_shop = true;
        product.price = Some(0);

        let (result, _) = request(&product).await;

        assert_eq!(
            result.errors.first().unwrap().message,
            "Product price must be greater than 0"
        );
    }

    #[tokio::test]
    pub async fn update() {
        let schema = async_graphql::Schema::build(Queries, Mutations, EmptySubscription).finish();
        let db = setup_db().await.unwrap();

        let mut product = ProductInput::mock();

        let request = make_request(&product).data(db.clone());

        // create
        schema.execute(request).await;

        product.title.en = "New title".into();
        let request = make_request(&product).data(db.clone());

        // update
        let result = schema.execute(request).await;

        assert_eq!(
            serde_json::to_string(&result).unwrap(),
            serde_json::json!({"data":{"saveProduct": {"id": product.id}}}).to_string()
        );

        let row = sqlx::query!("select * from products where id = ? ", product.id.0)
            .fetch_one(&db)
            .await
            .unwrap();

        assert_eq!(row.title_en, "New title");
    }
}
