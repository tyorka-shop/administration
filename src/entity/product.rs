use super::multi_lang::MultiLang;
use async_graphql::{Result, ID};
use serde::Serialize;

#[derive(macros::Entity)]
#[table_name = "products"]
pub struct Entity {
    pub id: String,
    pub cover_id: String,
    pub title_en: String,
    pub title_ru: String,
    pub description_en: String,
    pub description_ru: String,
    pub price: Option<i64>,
    pub show_in_gallery: bool,
    pub show_in_shop: bool,
}

#[derive(async_graphql::SimpleObject)]
#[graphql(complex)]
pub struct Product {
    pub id: ID,
    #[graphql(skip)]
    pub cover_id: String,
    pub title: MultiLang,
    pub show_in_gallery: bool,
    pub show_in_shop: bool,
    pub price: Option<i64>,
    pub description: MultiLang,
}

impl From<&Entity> for Product {
    fn from(row: &Entity) -> Self {
        Self {
            id: ID::from(&row.id),
            cover_id: row.cover_id.clone(),
            description: MultiLang {
                en: row.description_en.clone(),
                ru: row.description_ru.clone(),
            },
            price: row.price,
            show_in_gallery: row.show_in_gallery,
            show_in_shop: row.show_in_shop,
            title: MultiLang {
                en: row.title_en.clone(),
                ru: row.title_ru.clone(),
            },
        }
    }
}

#[derive(async_graphql::InputObject, Serialize, Clone)]
pub struct ProductInput {
    pub id: ID,
    pub pictures: Vec<ID>,
    pub cover_id: ID,
    pub title: MultiLang,
    pub show_in_gallery: bool,
    pub show_in_shop: bool,
    pub price: Option<i64>,
    pub description: MultiLang,
}

impl From<&ProductInput> for Entity {
    fn from(input: &ProductInput) -> Self {
        Self {
            id: input.id.to_string(),
            title_en: input.title.en.clone(),
            title_ru: input.title.ru.clone(),
            description_en: input.description.en.clone(),
            description_ru: input.description.ru.clone(),
            price: input.price,
            show_in_gallery: input.show_in_gallery,
            show_in_shop: input.show_in_shop,
            cover_id: input.cover_id.to_string(),
        }
    }
}

#[cfg(test)]
impl ProductInput {
    pub fn mock() -> Self {
        let pic_id = ID::from(format!("{:x}", md5::compute("cover_id")));
        Self {
            id: "07d7b72c-5b2e-4a35-a257-158496993dcc".into(),
            pictures: vec![pic_id.clone()],
            cover_id: pic_id,
            title: MultiLang {
                en: "title".to_string(),
                ru: "заголовок".to_string(),
            },
            show_in_gallery: true,
            show_in_shop: false,
            price: None,
            description: MultiLang {
                en: "description".to_string(),
                ru: "описание".to_string(),
            },
        }
    }
}

#[cfg(test)]
impl Entity {
    pub fn mock() -> Self {
        Self {
            id: "07d7b72c-5b2e-4a35-a257-158496993dcc".into(),
            cover_id: format!("{:x}", md5::compute("cover_id")),
            title_en: "title".to_string(),
            title_ru: "заголовок".to_string(),
            description_en: "description".to_string(),
            description_ru: "описание".to_string(),
            price: None,
            show_in_gallery: true,
            show_in_shop: false,
        }
    }
}
