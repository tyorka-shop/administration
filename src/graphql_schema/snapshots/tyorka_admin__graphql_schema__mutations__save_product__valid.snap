---
source: src/graphql_schema/mutations.rs
expression: result.data.into_json().unwrap()
---
{
  "saveProduct": {
    "id": "07d7b72c-5b2e-4a35-a257-158496993dcc"
  }
}
